cmd/ts/netcheck.go

```diff
@@ -0,0 +1,269 @@
+			log.Println("Failed to fetch a DERP map, so netcheck cannot continue. Check your Internet connection.")
+			return err
+		}
+	}
+	for {
+		t0 := time.Now()
+		report, err := c.GetReport(ctx, dm, nil)
+		d := time.Since(t0)
+		if cmd.verbose {
+			c.Logf("GetReport took %v; err=%v", d.Round(time.Millisecond), err)
+		}
+		if err != nil {
+			return fmt.Errorf("netcheck: %w", err)
+		}
+		if err := cmd.printReport(dm, report); err != nil {
+			return err
+		}
+		if cmd.every == 0 {
+			return nil
+		}
+		time.Sleep(cmd.every)
+	}
+}
+
+func (cmd *NetcheckCmd) printReport(dm *tailcfg.DERPMap, report *netcheck.Report) error {
+	var j []byte
+	var err error
+	switch cmd.format {
+	case "":
+	case "json":
+		j, err = json.MarshalIndent(report, "", "\t")
+	case "json-line":
+		j, err = json.Marshal(report)
+	default:
+		return fmt.Errorf("unknown output format %q", cmd.format)
+	}
+	if err != nil {
+		return err
+	}
+	if j != nil {
+		j = append(j, '\n')
+		Stdout.Write(j)
+		return nil
+	}
+
+	printf("\nReport:\n")
+	printf("\t* UDP: %v\n", report.UDP)
+	if report.GlobalV4.IsValid() {
+		printf("\t* IPv4: yes, %s\n", report.GlobalV4)
+	} else {
+		printf("\t* IPv4: (no addr found)\n")
+	}
+	if report.GlobalV6.IsValid() {
+		printf("\t* IPv6: yes, %s\n", report.GlobalV6)
+	} else if report.IPv6 {
+		printf("\t* IPv6: (no addr found)\n")
+	} else if report.OSHasIPv6 {
+		printf("\t* IPv6: no, but OS has support\n")
+	} else {
+		printf("\t* IPv6: no, unavailable in OS\n")
+	}
+	printf("\t* MappingVariesByDestIP: %v\n", report.MappingVariesByDestIP)
+	printf("\t* PortMapping: %v\n", portMapping(report))
+	if report.CaptivePortal != "" {
+		printf("\t* CaptivePortal: %v\n", report.CaptivePortal)
+	}
+
+	// When DERP latency checking failed,
+	// magicsock will try to pick the DERP server that
+	// most of your other nodes are also using
+	if len(report.RegionLatency) == 0 {
+		printf("\t* Nearest DERP: unknown (no response to latency probes)\n")
+	} else {
+		if report.PreferredDERP != 0 {
+			printf("\t* Nearest DERP: %v\n", dm.Regions[report.PreferredDERP].RegionName)
+		} else {
+			printf("\t* Nearest DERP: [none]\n")
+		}
+		printf("\t* DERP latency:\n")
+		var rids []int
+		for rid := range dm.Regions {
+			rids = append(rids, rid)
+		}
+		sort.Slice(rids, func(i, j int) bool {
+			l1, ok1 := report.RegionLatency[rids[i]]
+			l2, ok2 := report.RegionLatency[rids[j]]
+			if ok1 != ok2 {
+				return ok1 // defined things sort first
+			}
+			if !ok1 {
+				return rids[i] < rids[j]
+			}
+			return l1 < l2
+		})
+		for _, rid := range rids {
+			d, ok := report.RegionLatency[rid]
+			var latency string
+			if ok {
+				latency = d.Round(time.Millisecond / 10).String()
+			}
+			r := dm.Regions[rid]
+			var derpNum string
+			if cmd.verbose {
+				derpNum = fmt.Sprintf("derp%d, ", rid)
+			}
+			printf("\t\t- %3s: %-7s (%s%s)\n", r.RegionCode, latency, derpNum, r.RegionName)
+		}
+	}
+	return nil
+}
+
+func portMapping(r *netcheck.Report) string {
```

Comment:
can we move some of this stuff from `cmd/` to `pkg/tailscale`?

Reply:
I could do, are there functions here you think we could use elsewhere? The rationale here was that if we update our tailscale fork we will need to sync these commands with the upstream. So I thought the least refactoring I did the easier the updates would be. For instance most of these files are almost the exact same except for the first 50 lines, which are replaced by the cobra struct and injecting of the local client.

But if there are useful functions in here and you think it is worth us having this be more structured then happy to refactor like you said and move stuff to `pkg/tailscale` and clean up some of the harder to read code